csrc/preseg_passes/move_repeat_forward.cpp

```diff
@@ -0,0 +1,480 @@
+  }
+
+  const std::unordered_map<TensorView*, IterDomain*>& repeatIdMap() const {
+    return repeat_id_map_;
+  }
+
+ private:
+  // Mappings of tensors to their repeated logical IDs
+  std::unordered_map<TensorView*, IterDomain*> repeat_id_map_;
+  // Tvs that the repeat cannot be moved past
+  std::unordered_set<TensorView*> no_move_tvs_;
+};
+
+class MoveRepeatForward {
+  using StaticRepeatInfo = scheduler_tools::StaticRepeatInfo;
+
+ public:
+  MoveRepeatForward(Fusion* fusion) : fusion_(fusion) {}
+
+  void run() {
+    auto reshape_ops = ir_utils::getOpsOfType<ViewOp>(fusion_);
+    std::vector<TensorView*> reshape_output_tvs;
+    reshape_output_tvs.reserve(reshape_ops.size());
+    std::ranges::transform(
+        reshape_ops,
+        std::back_inserter(reshape_output_tvs),
+        [](ViewOp* reshape) { return reshape->out(); });
+
+    // For each reshape output, if it's a repeating reshape and a
+    // valid move target is found, try moving the repetition after the
+    // target
+    for (auto reshape_output_tv : reshape_output_tvs) {
+      auto static_repeat_info =
+          scheduler_tools::getMaybeStaticRepeatInfo(reshape_output_tv);
+      if (!static_repeat_info.has_value()) {
+        continue;
+      }
+
+      auto move_target_info = findMoveTarget(*static_repeat_info);
+      if (!move_target_info.has_value()) {
+        continue;
+      }
+
+      moveTo(
+          *static_repeat_info,
+          move_target_info->first,
+          move_target_info->second);
+    }
+  }
+
+ private:
+  // Try to find the furthest tensor that a given repetition can be
+  // moved after. Returns a tensor where the repetition should be
+  // moved if found. The map of repeat IDs is also returned as it is
+  // necessary for the later move step.
+  std::optional<
+      std::pair<TensorView*, std::unordered_map<TensorView*, IterDomain*>>>
+  findMoveTarget(const scheduler_tools::StaticRepeatInfo& info) {
+    auto reshape_output_tv = info.reshape_output_tv;
+    auto all_exprs = DependencyCheck::getAllExprsBetween(
+        {reshape_output_tv}, fusion_->outputs());
+
+    // Analyze all exprs after the repeating reshape. The returned map
+    // should have mappings if it's valid to move the repeat after the
+    // tensor.
+    const auto repeat_id_map =
+        CanMoveOver(reshape_output_tv, info.output_id, all_exprs).repeatIdMap();
+
+    // To find the furthest possible position, check the furthest post
+    // dominator and see if that's a valid position
+    const auto post_dominators = getAllPostDominators(reshape_output_tv);
+
+    TensorView* target_tv = nullptr;
+    for (const auto& post_dominator : post_dominators | std::views::reverse) {
+      NVF_ERROR(
+          post_dominator->isA<TensorView>(),
+          "Expected to be a tensor: ",
+          post_dominator->toString());
+
+      if (auto it = repeat_id_map.find(post_dominator->as<TensorView>());
+          it != repeat_id_map.end() && it->second != nullptr) {
```

Comment:
nitpick: should `it->second != nullptr` be an assert instead?

Reply:
Thanks. Originally, I was using `repeat_id_map` differently, and having a `nullptr` meant something else. Will change it to an assertion.